src/java.base/share/classes/jdk/internal/util/HexDigits.java

```diff
@@ -114,24 +114,18 @@ public static short digitPair(int i, boolean ucase) {
     }
 
     /**
-     * Return a little-endian packed integer for the 4 ASCII bytes for an input unsigned 2-byte integer.
-     * {@code b0} is the most significant byte and {@code b1} is the least significant byte.
-     * The integer is passed byte-wise to allow reordering of execution.
+     * Insert the int into the buffer as 4 hexadecimal digits
+     * @param buffer byte buffer to copy into
+     * @param off insert point
+     * @param value to convert
      */
-    public static int packDigits(int b0, int b1) {
-        return DIGITS[b0 & 0xff] | (DIGITS[b1 & 0xff] << 16);
-    }
-
-    /**
-     * Return a little-endian packed long for the 8 ASCII bytes for an input unsigned 4-byte integer.
-     * {@code b0} is the most significant byte and {@code b3} is the least significant byte.
-     * The integer is passed byte-wise to allow reordering of execution.
-     */
-    public static long packDigits(int b0, int b1, int b2, int b3) {
-        return DIGITS[b0 & 0xff]
-                | (DIGITS[b1 & 0xff] << 16)
-                | (((long) DIGITS[b2 & 0xff]) << 32)
-                | (((long) DIGITS[b3 & 0xff]) << 48);
+    public static void putHex(byte[] buffer, int off, int i) {
```

Comment:
Should there be 2 methods - for 2 and 4 bytes respectively?
Does c2 optimize 8 byte writes as well?

Reply:
The 4-byte unsigned int input for 8-byte write sounds plausible, I personally am fine either with or without it.

> Does c2 optimize 8 byte writes as well?

From the first few lines of #16245:

> Merging multiple consecutive small stores (e.g. 8 byte stores) into larger stores (e.g. one long store) can lead to speedup.

Reply:
8-byte writing requires converting int to long. The performance is similar to the current version, but an additional method putHex8 needs to be added. The current version has less code.

The following is the code for writing 8 bytes:
```java
class UUID {
     @Override
    public String toString() {
        byte[] buf = new byte[36];
        HexDigits.putHex8(buf, 0, (int) (mostSigBits >> 32));
        HexDigits.putHex10(buf, 8, (int) mostSigBits);
        HexDigits.putHex10(buf, 18, (int) (leastSigBits >> 32));
        HexDigits.putHex8(buf, 28, (int) leastSigBits);
        try {
            return jla.newStringNoRepl(buf, StandardCharsets.ISO_8859_1);
        } catch (CharacterCodingException cce) {
            throw new AssertionError(cce);
        }
    }
}

class HexDigits {
    public static void putHex8(byte[] bytes, int off, int i) {
        long v =  (((long) DIGITS[(i >> 16) & 0xff]) << 48)
                | (((long) DIGITS[(i >> 24) & 0xff]) << 32)
                | (        DIGITS[ i        & 0xff]  << 16)
                | (        DIGITS[(i >> 8 ) & 0xff]);
        bytes[off]     = (byte)  v;
        bytes[off + 1] = (byte) (v >> 8);
        bytes[off + 2] = (byte) (v >> 16);
        bytes[off + 3] = (byte) (v >> 24);
        bytes[off + 4] = (byte) (v >> 32);
        bytes[off + 5] = (byte) (v >> 40);
        bytes[off + 6] = (byte) (v >> 48);
        bytes[off + 7] = (byte) (v >> 56);
    }

    public static void putHex10(byte[] bytes, int off, int i) {
        int v0 = (DIGITS[(i >> 16) & 0xff] << 16)
                | DIGITS[(i >> 24) & 0xff];
        int v1 = (DIGITS[i         & 0xff] << 16)
                | DIGITS[(i >> 8 ) & 0xff];
        bytes[off]     = '-';
        bytes[off + 1] = (byte)  v0;
        bytes[off + 2] = (byte) (v0 >> 8);
        bytes[off + 3] = (byte) (v0 >> 16);
        bytes[off + 4] = (byte) (v0 >> 24);
        bytes[off + 5] = '-';
        bytes[off + 6] = (byte)  v1;
        bytes[off + 7] = (byte) (v1 >> 8);
        bytes[off + 8] = (byte) (v1 >> 16);
        bytes[off + 9] = (byte) (v1 >> 24);
    }
}
```